src/main/java/com/github/pgasync/impl/ConnectionValidator.java

```diff
@@ -26,26 +26,40 @@
 public class ConnectionValidator {
 
     final String validationQuery;
+    final boolean validateSocket;
 
-    public ConnectionValidator(String validationQuery) {
-        this.validationQuery = validationQuery;
+    public ConnectionValidator(String validationQuery, boolean validateSocket) {
+        // Trimmed as empty means no query for backwards compatibility
+        this.validationQuery = validationQuery == null || validationQuery.trim().isEmpty() ? null : validationQuery;
+        this.validateSocket = validateSocket;
     }
 
     public Observable<Connection> validate(Connection connection) {
-        return connection.queryRows(validationQuery)
-                .lift(subscriber -> new Subscriber<Row>() {
-                    @Override
-                    public void onError(Throwable e) {
-                        subscriber.onError(e);
-                    }
-                    @Override
-                    public void onCompleted() {
-                        subscriber.onNext(connection);
-                        subscriber.onCompleted();
-                    }
-                    @Override
-                    public void onNext(Row row) { }
-                });
+        Observable<Connection> ret = Observable.just(connection);
+        if (validationQuery != null) {
+            ret = ret.flatMap(conn -> connection.queryRows(validationQuery)
+                    .lift(subscriber -> new Subscriber<Row>() {
+                        @Override
+                        public void onError(Throwable e) {
+                            subscriber.onError(e);
+                        }
+                        @Override
+                        public void onCompleted() {
+                            subscriber.onNext(connection);
+                            subscriber.onCompleted();
+                        }
+                        @Override
+                        public void onNext(Row row) { }
+                    }));
+        }
+        if (validateSocket) {
+            ret = ret.doOnNext(conn -> {
+                if (conn instanceof PgConnection && !((PgConnection) conn).isConnected()) {
+                    throw new IllegalStateException("Channel is closed");
+                }
+            });
+        }
+        return ret;
```

Comment:
it looks like you got two validation in this file - consider separating them, they do not need to have fields - just function with params.

Also I wander if you need the second one - when you are able to query does it mean you are able to connect?

Reply:
In my case, I didn't have a validation query, only a socket validation. They are separate, but I figured placing all forms of validation in `ConnectionValidator` was reasonable since it's the validation result that determines whether it needs to be refreshed regardless of the method you choose to validate. I was avoiding an over-abstraction here but we could easily abstract out many ways to validate a connection.

Reply:
It just looks like a two separate way of doing validation. Possibly one could have an option to use each one of them etc. This is just general observation about it.

Reply:
You should firstly check if there is socket connection then if you can query.

Reply:
Basically this socket check doesn't make much sense as if it is closed then the check query will fail.

Reply:
@jaceksokol - I explained this above with "I didn't have a validation query, only a socket validation." I don't want query validation. The socket dies underneath and I don't need a full query to the DB for the client side to detect socket death. People should not have to run queries against the DB to see if the client-side of the socket has died.

I just want the client-side socket death check (without requiring a validation query) and I figured the connection validator is the same place. I put it after because I don't want to change anything for how people rely on things today, just need to tack it on.